dashboard/utils.py

```diff
@@ -0,0 +1,180 @@
+"""
+Utility functions and classes for the dashboard
+"""
+from django.db.models import Max, Min, Q
+
+from courses.models import CourseRun
+from ecommerce.models import Line
+from financialaid.models import FinancialAid, TierProgram
+
+
+class RequiredFieldMissing(Exception):
+    """
+    Custom exception to raise in case a field that is not mandatory in the db
+    is mandatory for the implementation of some business logic.
+    Example CourseRun end date.
+    """
+
+
+class MMTrack:
+    """
+    Abstraction around the user status in courses.
+    Needed because the user enrollment "verified" and user passed status
+    can be checked in different ways depending on if the program offers financial aid or not.
+    """
+    # pylint: disable=too-many-instance-attributes, too-many-arguments
+
+    user = None
+    program = None
+    enrollments = None
+    current_grades = None
+    certificates = None
+    course_ids = []
+    paid_course_ids = []
+    financial_aid_available = None
+    financial_aid_applied = None
+    financial_aid_status = None
+    financial_aid = None
+    financial_aid_min_price = None
+    financial_aid_max_price = None
+
+    def __init__(self, user, program, enrollments, current_grades, certificates):
+        """
+        Args:
+            user (User): a Django user
+            program (programs.models.Program): program where the user is enrolled
+            enrollments (edx_api.enrollment.models.Enrollments): User's enrollments
+            current_grades (edx_api.grades.models.CurrentGrades): User's current grades
+            certificates (edx_api.certificates.models.Certificates): User's certificates
+        """
+        self.user = user
+        self.program = program
+        self.enrollments = enrollments
+        self.current_grades = current_grades
+        self.certificates = certificates
+        self.financial_aid_available = program.financial_aid_availability
+
+        self.course_ids = list(CourseRun.objects.filter(course__program=self.program).exclude(
+            Q(edx_course_key__isnull=True) | Q(edx_course_key__exact='')
+        ).values_list("edx_course_key", flat=True))
+
+        if self.financial_aid_available:
+            self.paid_course_ids = list(Line.objects.filter(
```

Comment:
I think I understand now, `paid_course_ids` is defined to only include payments on our own servers, right?